### ISSUE TITLE
can anybody tried FP 16 training?

### ISSUE THREAD

**Issue by dedoogong** [tianzhi0549/FCOS#121]
I'm trying very hard to train FCOS on FP16 but I got many similar errors showing data type conflict between expected float type and actually converted half type. 

Using apex, I modified small parts as below

1. [trainer.py]
// losses.backward() is replaced by 
with amp.scale_loss(losses, optimizer) as scaled_loss:
      scaled_loss.backward()

2. [sigmoid_focal_loss.py]
// from ctx.save_for_backward(logits, targets) to 
ctx.save_for_backward(logits.float(), targets.int()) 

after many different trials for changing float<->half casing on input/output for model or loss arguments, I got just similar type casting errors without any clever solution...

Please let me know any hint if you already tried FP 16 training. 

Thank you!

**Comment by tianzhi0549** [comment]
@dedoogong If the major problem is `NAN`, can you try to clip the gradients?

**Comment by dedoogong** [comment]
yes I'm trying to apply @amp.float() annotation for the sigmoid focal loss's forward pass and I can run training now using only 1 GPU(batch 8). if I run multi gpu training, the grad.type()==variable.type() assert error occurred.

**Comment by tianzhi0549** [comment]
@dedoogong I am not sure but have you tried to use `apex.DistributedDataParallel` instead of the one in torch?

**Comment by dedoogong** [comment]
@tianzhi0549 yes I can run now with your advice. but with O1 optimize option, 
FCOS-MVN2-FPN-ms-c128 model's total loss goes down to just 2.3~2.4
and, instead of MVN2, after replacing my custom mobilenet-like backbone, it also goes down to 2.3~2.4..I will try O2 optimization, but hum...I'm not sure why accuracy is so low like that.

**Comment by tianzhi0549** [comment]
@dedoogong I am not sure either. Maybe you need to check some details.

**Comment by dedoogong** [comment]
@tianzhi0549 yes me too. after changing LR to 1.0 with steps[5000, ... ], it shows slightly better loss(around 2.1xxx) but that's all. Now I'm trying to just convert your pretrained FCOS-MVN2-FPN-ms-c128 model to fp16 using model.half() with input--images.half(). I hope to see over 40% speed-up and little less mAP.

**Comment by stanstarks** [comment]
Hi @dedoogong ,
Just replied to you under the maskrcnn project with my workaround.

https://github.com/facebookresearch/maskrcnn-benchmark/issues/1048#issuecomment-526448630

However, I don't have enough resources to test its performance. Could you try and tell us how it goes?

Cheers.

**Comment by dedoogong** [comment]
@stanstarks 
Thank you for your reply, 
that workaround didn't work for apex's DDP. it worked for only torch's DDP. so I needed to use apexes DDP and used @float type annotator(sorry I didnt remember the exact annotator name) for the forward function of the sigmoid focal loss class. 

Anyway, My custom backbone showed better mAP than MVN2, even higher than MVN3. But FCOS-MyBackbone-FPN-C128-ms's mAP is just 25%-- this is the best result so far. I have trained my model using so many various hyper parameters. And I got that result with batch size of 16. With bs32, it was just around 20%.
 
When I trained FCOS-bs32-MVN2-FPN-c128-ms version on my DGX-1, I got 30.4%
And when I trained both my custom version and MVN2 version on FP16, both gave so bad result(less than 10%).

I'm still trying to get the better mAP.
